icc-x-api/icc-crypto-x-api.ts

```diff
@@ -1358,39 +1347,154 @@ export class IccCryptoXApi {
     }
   }
 
-  saveKeyChainInHCPFromLocalStorage(hcpId: string): Promise<HealthcarePartyDto> {
-    return this.hcpartyBaseApi
+  /**
+   * Save eHealth certificate and validity in the encrypted fields of the HCP.
+   * @param hcpId Id of the hcp to modify
+   * @param autoupdateHCP update the HCP using the base HCP service; if false simply return the new HCP,
+   * without publishing the modifications on the db (this way this can be handled elsewhere e.g. by the
+   * HCP x api service, which makes use of caching).
+   */
+  async saveKeyChainInHCPFromLocalStorage(
+    hcpId: string,
+    autoupdateHCP = true
+  ): Promise<HealthcarePartyDto> {
+    return await this.hcpartyBaseApi
       .getHealthcareParty(hcpId)
-      .then(hcp => {
+      .then(async (hcp: HealthcarePartyDto) => {
+        let aesKey: CryptoKey | null = null
+        try {
+          aesKey = _.find(
+            await this.decryptAndImportAesHcPartyKeysForDelegators([hcp.id!], hcp.id!),
+            delegator => delegator.delegatorId === hcp.id
+          )!.key
+        } catch (e) {
+          console.error("Error while importing the AES key.")
+        }
+        if (!aesKey) {
+          console.error("No encryption key!")
+        }
+
         const opts = hcp.options || {}
 
         const crt = this.getKeychainInBrowserLocalStorageAsBase64(hcp.id!!)
         _.set(opts, this.hcpPreferenceKeyEhealthCert, crt)
 
+        if (!!aesKey && !!crt) {
+          let crtEncrypted: ArrayBuffer | null = null
+          try {
+            crtEncrypted = await this.AES.encrypt(aesKey, new Uint8Array(utils.text2ua(atob(crt))))
+          } catch (e) {
+            console.error("Error while encrypting the certificate", e)
+          }
+          _.set(
+            opts,
+            this.hcpPreferenceKeyEhealthCertEncrypted,
+            utils.ua2text(new Uint8Array(crtEncrypted!))
+          )
+        }
+
         const crtValidityDate = this.getKeychainValidityDateInBrowserLocalStorage(hcp.id!!)
         if (!!crtValidityDate) {
           _.set(opts, this.hcpPreferenceKeyEhealthCertDate, crtValidityDate)
```

Comment:
The unencrypted data is still stored here. Instead we should clear it if present.

Reply:
https://github.com/medispring/icc-api/pull/193/files#r467850334

Reply:
Done!